sound/soc/codecs/rt711-sdca-sdw.c

```diff
@@ -0,0 +1,479 @@
+	case SDW_SDCA_CTL(FUN_MIC_ARRAY, RT711_SDCA_ENT_USER_FU1E, RT711_SDCA_CTL_FU_VOLUME, CH_L):
+	case SDW_SDCA_CTL(FUN_MIC_ARRAY, RT711_SDCA_ENT_USER_FU1E, RT711_SDCA_CTL_FU_VOLUME, CH_R):
+	case SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_USER_FU0F, RT711_SDCA_CTL_FU_VOLUME, CH_L):
+	case SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_USER_FU0F, RT711_SDCA_CTL_FU_VOLUME, CH_R):
+	case SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_PLATFORM_FU44, RT711_SDCA_CTL_FU_CH_GAIN, CH_L):
+	case SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_PLATFORM_FU44, RT711_SDCA_CTL_FU_CH_GAIN, CH_R):
+	case SDW_SDCA_CTL(FUN_MIC_ARRAY, RT711_SDCA_ENT_PLATFORM_FU15, RT711_SDCA_CTL_FU_CH_GAIN, CH_L):
+	case SDW_SDCA_CTL(FUN_MIC_ARRAY, RT711_SDCA_ENT_PLATFORM_FU15, RT711_SDCA_CTL_FU_CH_GAIN, CH_R):
+		ret = regmap_read(rt711->sdw_regmap, reg, &sdw_data_0);
+		if (ret < 0)
+			return ret;
+		ret = regmap_read(rt711->sdw_regmap, SDW_SDCA_MBQ_CTL(reg), &sdw_data_1);
+		if (ret < 0)
+			return ret;
+		*val = ((sdw_data_1 & 0xff) << 8) | (sdw_data_0 & 0xff);
+		break;
+	default:
+		/* SDCA mapping case or normal address */
+		if (reg & 0x40000000 || reg < 0xffff) {
+			ret = regmap_read(rt711->sdw_regmap, reg, val);
+			if (ret < 0)
+				return ret;
+		}
+
+		/* vendor registers case */
+		if (reg < 0x40000000 && reg > 0xffff) {
+			ret = regmap_read(rt711->sdw_regmap, reg, &sdw_data_0);
+			if (ret < 0)
+				return ret;
+			ret = regmap_read(rt711->sdw_regmap,
+				SDW_SDCA_MBQ_CTL(reg), &sdw_data_1);
+			if (ret < 0)
+				return ret;
+			*val = ((sdw_data_1 & 0xff) << 8) | (sdw_data_0 & 0xff);
+		}
+		break;
+	}
+
+	dev_dbg(dev, "[%s] %04x => %08x\n", __func__, reg, *val);
+	return 0;
+}
+
+static int rt711_sdca_sdw_write(void *context, unsigned int reg, unsigned int val)
+{
+	struct device *dev = context;
+	struct rt711_sdca_priv *rt711 = dev_get_drvdata(dev);
+
+	switch (reg) {
+	case SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_USER_FU05, RT711_SDCA_CTL_FU_VOLUME, CH_L):
+	case SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_USER_FU05, RT711_SDCA_CTL_FU_VOLUME, CH_R):
+	case SDW_SDCA_CTL(FUN_MIC_ARRAY, RT711_SDCA_ENT_USER_FU1E, RT711_SDCA_CTL_FU_VOLUME, CH_L):
+	case SDW_SDCA_CTL(FUN_MIC_ARRAY, RT711_SDCA_ENT_USER_FU1E, RT711_SDCA_CTL_FU_VOLUME, CH_R):
+	case SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_USER_FU0F, RT711_SDCA_CTL_FU_VOLUME, CH_L):
+	case SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_USER_FU0F, RT711_SDCA_CTL_FU_VOLUME, CH_R):
+	case SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_PLATFORM_FU44, RT711_SDCA_CTL_FU_CH_GAIN, CH_L):
+	case SDW_SDCA_CTL(FUN_JACK_CODEC, RT711_SDCA_ENT_PLATFORM_FU44, RT711_SDCA_CTL_FU_CH_GAIN, CH_R):
+	case SDW_SDCA_CTL(FUN_MIC_ARRAY, RT711_SDCA_ENT_PLATFORM_FU15, RT711_SDCA_CTL_FU_CH_GAIN, CH_L):
+	case SDW_SDCA_CTL(FUN_MIC_ARRAY, RT711_SDCA_ENT_PLATFORM_FU15, RT711_SDCA_CTL_FU_CH_GAIN, CH_R):
+		regmap_write(rt711->sdw_regmap, SDW_SDCA_MBQ_CTL(reg), ((val >> 8) & 0xff));
+		regmap_write(rt711->sdw_regmap, reg, (val & 0xff));
```

Comment:
@shumingfan I wonder if this solution actually works for the suspend/resume parts, where we use regmap_sync() functions to restore the state. Here you manually issue two regmap_write(), but on a resume the sync() would not use the MBQ bit, so would all these volumes be restored after a system suspend where the power might be turned off?

Reply:
See suggested solution in https://github.com/thesofproject/linux/pull/2364

Reply:
The regcache_sync() will restore the settings, let me show you.
The regmap_config (rt711_sdca_regmap) has own .reg_write/.reg_read callback funcitons.
The codec driver will use this register map to access registers.
The callback functions will check the register address to decide the access method using MBQ bit or not.

There are test logs below. The regcache_sync() operation can restore the settings.
root@Ice-Lake:~# amixer -c0 cset numid=7 87
numid=7,iface=MIXER,name='rt711 FU05 Playback Volume'
  ; type=INTEGER,access=rw---R--,values=2,min=0,max=87,step=0
  : values=87,87
  | dBscale-min=-65.25dB,step=0.75dB,mute=0

[  204.388758] rt711-sdca sdw:2:25d:711:1: 40402291 <= 0
[  204.389198] rt711-sdca sdw:2:25d:711:1: 40400291 <= 0
[  204.389612] rt711-sdca sdw:2:25d:711:1: [rt711_sdca_sdw_write] 40400291 <= 0000
[  204.389614] rt711-sdca sdw:2:25d:711:1: 40402292 <= 0
[  204.390045] rt711-sdca sdw:2:25d:711:1: 40400292 <= 0
[  204.390597] rt711-sdca sdw:2:25d:711:1: [rt711_sdca_sdw_write] 40400292 <= 0000 
 
After the run-time suspend, I changed the volume settings.
root@Ice-Lake:~# amixer -c0 cset numid=7 50
numid=7,iface=MIXER,name='rt711 FU05 Playback Volume'
  ; type=INTEGER,access=rw---R--,values=2,min=0,max=87,step=0
  : values=50,50
  | dBscale-min=-65.25dB,step=0.75dB,mute=0
  
[  371.980822] rt711_sdca_dev_resume, start to reg sync
...
[  372.009651] rt711-sdca sdw:2:25d:711:1: 40400291 <= e440
[  372.009651] rt711-sdca sdw:2:25d:711:1: 4040**2**291 <= e4
[  372.010070] rt711-sdca sdw:2:25d:711:1: 40400291 <= 40
[  372.010484] rt711-sdca sdw:2:25d:711:1: [rt711_sdca_sdw_write] 40400291 <= e440
[  372.010485] rt711-sdca sdw:2:25d:711:1: 40400292 <= e440
[  372.010486] rt711-sdca sdw:2:25d:711:1: 4040**2**292 <= e4
[  372.010859] rt711-sdca sdw:2:25d:711:1: 40400292 <= 40
[  372.011290] rt711-sdca sdw:2:25d:711:1: [rt711_sdca_sdw_write] 40400292 <= e440
...
[  372.021935] rt711_sdca_dev_resume, reg sync done

Reply:
My point was that it forces you to declare each control twice, and then have a test on the address to force the second byte to be updated - essentially preventing the second byte from actually being managed by regmap.

it's really a manual way of managing these registers and it makes every codec driver a bit more complicated than it needs to be.

Reply:
I agree with Pierre. We should support MBQ on regmap. It is not only for Realtek codecs but also for all codecs which support SDCA.

Reply:
@plbossart @bardliao Sure, I will update the codec driver and test with #2364.